x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -337,107 +348,163 @@ private RestoreSession openSession(String repositoryName, Client remoteClient, S
+                final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+                final AtomicReference<Tuple<StoreFileMetaData, Exception>> error = new AtomicReference<>();
+
+                final ArrayDeque<FileInfo> remainingFiles = new ArrayDeque<>(filesToRecover);
+                final ArrayDeque<FileInfo> inFlightRequests = new ArrayDeque<>();
+                final Object mutex = new Object();
+
+                while (true) {
+                    if (error.get() != null) {
+                        break;
+                    }
+                    final FileInfo fileToRecover;
+                    final long waitForOp;
+                    synchronized (mutex) {
+                        if (inFlightRequests.isEmpty() && remainingFiles.isEmpty()) {
+                            break;
+                        }
+                        final long maxConcurrentFileChunks = ccrSettings.getMaxConcurrentFileChunks();
+                        if (remainingFiles.isEmpty() == false && inFlightRequests.size() < maxConcurrentFileChunks) {
+                            for (int i = 0; i < maxConcurrentFileChunks; i++) {
+                                if (remainingFiles.isEmpty()) {
+                                    break;
+                                }
+                                inFlightRequests.push(remainingFiles.pop());
+                            }
+                        }
+                        fileToRecover = inFlightRequests.removeFirst();
+                        inFlightRequests.addLast(fileToRecover);
+                        waitForOp = requestSeqIdTracker.getMaxSeqNo() + 1 - inFlightRequests.size();
+                    }
+                    try {
+                        requestSeqIdTracker.waitForOpsToComplete(waitForOp);
+                        synchronized (mutex) {
+                            // if file has been removed in the mean-while, it means that restore of this file completed, so start working
+                            // on the next one
+                            if (inFlightRequests.contains(fileToRecover) == false) {
+                                continue;
+                            }
+                        }
+                        final int bytesRequested = Math.toIntExact(Math.min(ccrSettings.getChunkSize().getBytes(), fileToRecover.length()));
+                        final GetCcrRestoreFileChunkRequest request =
+                            new GetCcrRestoreFileChunkRequest(node, sessionUUID, fileToRecover.name(), bytesRequested);
+                        logger.trace("[{}] [{}] fetching chunk for file [{}]", shardId, snapshotId, fileToRecover.name());
+
+                        final long requestSeqId = requestSeqIdTracker.generateSeqNo();
+                        try {
+                            remoteClient.execute(GetCcrRestoreFileChunkAction.INSTANCE, request,
```

Comment:
We have lost the timeout. You can configure request options by overriding a method on the action.

Unfortunately it is not apparent to me if there is a way to get them back while using callbacks. We could always move the restore process into a local transport action and use the guice injections to get the `RemoteClusterService` and `TransportService`. But that is probably beyond the scope of this PR.

Reply:
yeah, I would like us to investigate this in a follow-up.